File: DentalWave/src/main/java/com/summerproject2026/DentalWave/service/ScheduleService.java
package com.summerproject2026.DentalWave.service;

import com.summerproject2026.DentalWave.dto.ScheduleDto;

import java.time.LocalDate;
import java.util.List;

/**
 * Service interface defining all business operations for Schedule management.
 * Includes CRUD, filtering, team assignment, and publishing.
 */
public interface ScheduleService {

    /**
     * Creates a new standalone schedule (not linked to a calendar at creation time).
     * Use CalendarService#addSchedule to create a schedule inside a calendar.
     *
     * @param scheduleDto the data for the new schedule
     * @return the persisted ScheduleDto
     */
    ScheduleDto createSchedule(ScheduleDto scheduleDto);

    /**
     * Retrieves a schedule by its ID.
     *
     * @param id the schedule ID
     * @return the matching ScheduleDto
     * @throws com.summerproject2026.DentalWave.exception.ResourceNotFoundException if not found
     */
    ScheduleDto getScheduleById(Long id);

    /**
     * Returns all schedules in the system.
     *
     * @return list of all ScheduleDtos (may be empty)
     */
    List<ScheduleDto> getAllSchedules();

    /**
     * Updates an existing schedule with new field values.
     *
     * @param id          the ID of the schedule to update
     * @param scheduleDto the updated data
     * @return the updated ScheduleDto
     * @throws com.summerproject2026.DentalWave.exception.ResourceNotFoundException if not found
     */
    ScheduleDto updateSchedule(Long id, ScheduleDto scheduleDto);

    /**
     * Deletes a schedule by ID.
     *
     * @param id the ID of the schedule to delete
     * @throws com.summerproject2026.DentalWave.exception.ResourceNotFoundException if not found
     */
    void deleteSchedule(Long id);

    /**
     * Returns all schedules that fall on a specific date.
     *
     * @param date the date to filter by
     * @return list of matching ScheduleDtos
     */
    List<ScheduleDto> getSchedulesByDate(LocalDate date);

    /**
     * Returns all schedules belonging to a given calendar.
     *
     * @param calendarId the parent calendar ID
     * @return list of matching ScheduleDtos
     */
    List<ScheduleDto> getSchedulesByCalendar(Long calendarId);

    /**
     * Returns all published schedules assigned to a given employee by ID.
     * Used by UC2 — Employee Views Personal Calendar.
     *
     * @param employeeId the employee ID
     * @return list of published ScheduleDtos assigned to the employee
     * @throws com.summerproject2026.DentalWave.exception.ResourceNotFoundException if employee not found
     */
    List<ScheduleDto> getSchedulesByEmployee(Long employeeId);

    /**
     * Returns all published schedules assigned to a given employee by full name.
     * Used by UC2 — Employee Views Personal Calendar.
     * Name matching is case-insensitive e.g. "jane smith" matches "Jane Smith".
     *
     * @param employeeName the full name of the employee e.g. "Jane Smith"
     * @return list of published ScheduleDtos assigned to the employee
     */
    List<ScheduleDto> getSchedulesByEmployeeName(String employeeName);

    /**
     * Assigns an employee to a specific team lead's team for this schedule.
     * If the user key doesn't exist in the teams map yet, it is created.
     *
     * @param scheduleId the schedule to update
     * @param userId     the ID of the team lead (User)
     * @param employeeId the ID of the Employee to add
     * @return the updated ScheduleDto
     * @throws com.summerproject2026.DentalWave.exception.ResourceNotFoundException if any entity is not found
     */
    ScheduleDto assignEmployeeToTeam(Long scheduleId, Long userId, Long employeeId);

    /**
     * Removes an employee from a specific team lead's team for this schedule.
     *
     * @param scheduleId the schedule to update
     * @param userId     the ID of the team lead (User)
     * @param employeeId the ID of the Employee to remove
     * @return the updated ScheduleDto
     * @throws com.summerproject2026.DentalWave.exception.ResourceNotFoundException if any entity is not found
     */
    ScheduleDto removeEmployeeFromTeam(Long scheduleId, Long userId, Long employeeId);

    /**
     * Marks a schedule as published.
     *
     * @param id the schedule ID
     * @return the updated ScheduleDto
     * @throws com.summerproject2026.DentalWave.exception.ResourceNotFoundException if not found
     */
    ScheduleDto publishSchedule(Long id);
}

File: DentalWave-frontend/src/services/ScheduleService.js
import axios from 'axios'
import { getAuthHeader } from './AuthService'

/**
 * Base URL for Schedule endpoints.
 */
const SCHEDULE_REST_API_BASE_URL = 'http://localhost:8080/api/schedules'

/**
 * Creates a new schedule.
 *
 * @param schedule the schedule data to create
 * @returns {Promise<axios.AxiosResponse<any>>}
 */
export const createSchedule = (schedule) =>
    axios.post(SCHEDULE_REST_API_BASE_URL, schedule, getAuthHeader())

/**
 * Gets a single schedule by id.
 *
 * @param scheduleId the schedule id
 * @returns {Promise<axios.AxiosResponse<any>>}
 */
export const getScheduleById = (scheduleId) =>
    axios.get(`${SCHEDULE_REST_API_BASE_URL}/${scheduleId}`, getAuthHeader())

/**
 * Gets all schedules.
 *
 * @returns {Promise<axios.AxiosResponse<any>>}
 */
export const getAllSchedules = () =>
    axios.get(SCHEDULE_REST_API_BASE_URL, getAuthHeader())

/**
 * Updates an existing schedule.
 *
 * @param scheduleId the schedule id
 * @param schedule the updated schedule data
 * @returns {Promise<axios.AxiosResponse<any>>}
 */
export const updateSchedule = (scheduleId, schedule) =>
    axios.put(`${SCHEDULE_REST_API_BASE_URL}/${scheduleId}`, schedule, getAuthHeader())

/**
 * Deletes a schedule.
 *
 * @param scheduleId the schedule id
 * @returns {Promise<axios.AxiosResponse<any>>}
 */
export const deleteSchedule = (scheduleId) =>
    axios.delete(`${SCHEDULE_REST_API_BASE_URL}/${scheduleId}`, getAuthHeader())

/**
 * Gets all schedules for a specific date.
 *
 * @param date ISO date string e.g. '2026-06-29'
 * @returns {Promise<axios.AxiosResponse<any>>}
 */
export const getSchedulesByDate = (date) =>
    axios.get(`${SCHEDULE_REST_API_BASE_URL}/date/${date}`, getAuthHeader())

/**
 * Gets all schedules for a specific calendar.
 *
 * @param calendarId the calendar id
 * @returns {Promise<axios.AxiosResponse<any>>}
 */
export const getSchedulesByCalendar = (calendarId) =>
    axios.get(`${SCHEDULE_REST_API_BASE_URL}/calendar/${calendarId}`, getAuthHeader())

/**
 * Gets all published schedules for a specific employee.
 * Used by UC2 — Employee Views Personal Calendar.
 *
 * @param employeeId the employee id
 * @returns {Promise<axios.AxiosResponse<any>>}
 */
export const getSchedulesByEmployee = (employeeId) =>
    axios.get(`${SCHEDULE_REST_API_BASE_URL}/employee/${employeeId}`, getAuthHeader())

/**
 * Gets all published schedules for a specific employee by name.
 * Used by UC2 — Employee Views Personal Calendar.
 *
 * @param employeeName the employee's full name
 * @returns {Promise<axios.AxiosResponse<any>>}
 */
export const getSchedulesByEmployeeName = (employeeName) =>
    axios.get(`${SCHEDULE_REST_API_BASE_URL}/employee/name/${employeeName}`, getAuthHeader())

/**
 * Assigns an employee to a specific team within a schedule.
 *
 * @param scheduleId the schedule the team belongs to
 * @param teamId the id of the team to add the employee to
 * @param employeeId the employee to add
 * @returns {Promise<axios.AxiosResponse<any>>}
 */
export const assignEmployeeToTeam = (scheduleId, teamId, employeeId) =>
    axios.post(
        `${SCHEDULE_REST_API_BASE_URL}/${scheduleId}/teams/${teamId}/employees/${employeeId}`,
        null,
        getAuthHeader()
    )

/**
 * Removes an employee from a specific team within a schedule.
 *
 * @param scheduleId the schedule the team belongs to
 * @param teamId the id of the team to remove the employee from
 * @param employeeId the employee to remove
 * @returns {Promise<axios.AxiosResponse<any>>}
 */
export const removeEmployeeFromTeam = (scheduleId, teamId, employeeId) =>
    axios.delete(
        `${SCHEDULE_REST_API_BASE_URL}/${scheduleId}/teams/${teamId}/employees/${employeeId}`,
        getAuthHeader()
    )

export const assignResourceToTeam = (scheduleId, teamId, resourceId) =>
    axios.post(
        `${SCHEDULE_REST_API_BASE_URL}/${scheduleId}/teams/${teamId}/resources/${resourceId}`,
        null,
        getAuthHeader()
    )

export const removeResourceFromTeam = (scheduleId, teamId, resourceId) =>
    axios.delete(
        `${SCHEDULE_REST_API_BASE_URL}/${scheduleId}/teams/${teamId}/resources/${resourceId}`,
        getAuthHeader()
    )

export const updateEmployeePartialDayNote = (scheduleId, teamId, employeeId, note) =>
    axios.put(
        `${SCHEDULE_REST_API_BASE_URL}/${scheduleId}/teams/${teamId}/employees/${employeeId}/partial-day`,
        { note },
        getAuthHeader()
    )

export const updateResourcePartialDayNote = (scheduleId, teamId, resourceId, note) =>
    axios.put(
        `${SCHEDULE_REST_API_BASE_URL}/${scheduleId}/teams/${teamId}/resources/${resourceId}/partial-day`,
        { note },
        getAuthHeader()
    )

/**
 * Publishes a schedule.
 *
 * @param scheduleId the schedule id
 * @returns {Promise<axios.AxiosResponse<any>>}
 */
export const publishSchedule = (scheduleId) =>
    axios.patch(`${SCHEDULE_REST_API_BASE_URL}/${scheduleId}/publish`, null, getAuthHeader())

/**
 * Creates a named team for a schedule.
 *
 * @param scheduleId the schedule id
 * @param name the team label/name
 * @returns {Promise<axios.AxiosResponse<any>>}
 */
export const createTeam = (scheduleId, name) =>
    axios.post(
        'http://localhost:8080/api/schedule-teams',
        { name, scheduleId, employees: [] },
        getAuthHeader()
    )

/**
 * Renames a schedule team.
 *
 * @param teamId the team id
 * @param name the new name
 * @returns {Promise<axios.AxiosResponse<any>>}
 */
export const renameTeam = (teamId, name) =>
    axios.put(
        `http://localhost:8080/api/schedule-teams/${teamId}/name`,
        null,
        { ...getAuthHeader(), params: { name } }
    )

/**
 * Deletes a schedule team.
 *
 * @param teamId the team id
 * @returns {Promise<axios.AxiosResponse<any>>}
 */
export const deleteTeam = (teamId) =>
    axios.delete(`http://localhost:8080/api/schedule-teams/${teamId}`, getAuthHeader())
